File: src/Test/StateMonadFFI.go
package Test_StateMonadFFI

type stateResult[S, A any] struct {
	val   A
	state S
}

type State[S, A any] func(S) stateResult[S, A]

func runState[S, A any](action State[S, A], initial S) stateResult[S, A] {
	return action(initial)
}

func bindState[S, A, B any](action State[S, A], next func(A) State[S, B]) State[S, B] {
	return func(state S) stateResult[S, B] {
		first := action(state)
		return next(first.val)(first.state)
	}
}

func pureState[S, A any](value A) State[S, A] {
	return func(state S) stateResult[S, A] { return stateResult[S, A]{val: value, state: state} }
}

func get[S any]() State[S, S] {
	return func(state S) stateResult[S, S] { return stateResult[S, S]{val: state, state: state} }
}

func put[S any](state S) State[S, struct{}] {
	return func(_ S) stateResult[S, struct{}] { return stateResult[S, struct{}]{val: struct{}{}, state: state} }
}

func modify[S any](f func(S) S) State[S, struct{}] {
	return bindState(get[S](), func(state S) State[S, struct{}] { return put(f(state)) })
}

func chainModifications(n int) State[int, struct{}] {
	if n == 0 {
		return pureState[int](struct{}{})
	}
	return bindState(modify(func(x int) int { return x + 1 }), func(_ struct{}) State[int, struct{}] {
		return chainModifications(n - 1)
	})
}

func runManyTimes_StateMonad(n int, depth int, acc int) int {
	if n == 0 {
		return acc
	}
	return runManyTimes_StateMonad(n-1, depth, acc+runState(chainModifications(depth), 0).state)
}

func RunStateMonadFFI(limit int) int {
	// The shared wrapper supplies depth 60; PureScript performs 20 repetitions.
	return runManyTimes_StateMonad(20, limit, 0)
}

File: src/Test/StateMonadFFI.java
    private record StateResult<S, A>(A value, S state) {}

    @FunctionalInterface
    private interface State<S, A> {
        StateResult<S, A> run(S initial);
    }

    private static <S, A, B> State<S, B> bindState(State<S, A> action, java.util.function.Function<A, State<S, B>> next) {
        return initial -> {
            StateResult<S, A> first = action.run(initial);
            return next.apply(first.value()).run(first.state());
        };
    }

    private static <S, A> State<S, A> pureState(A value) {
        return initial -> new StateResult<>(value, initial);
    }

    private static <S> State<S, S> getState() {
        return initial -> new StateResult<>(initial, initial);
    }

    private static <S> State<S, Void> putState(S value) {
        return ignored -> new StateResult<>(null, value);
    }

    private static <S> State<S, Void> modifyState(java.util.function.Function<S, S> update) {
        State<S, S> get = getState();
        return bindState(get, current -> putState(update.apply(current)));
    }

    private static State<Integer, Void> chainModifications(int depth) {
        if (depth == 0) return pureState(null);
        return bindState(modifyState(value -> value + 1), ignored -> chainModifications(depth - 1));
    }

    public static final java.util.function.Function<Object, Object> runStateMonadFFI = Bench.nativeBenchmark(input -> {
        // The shared FFI wrapper supplies depth 60; the PureScript test supplies 20 repetitions.
        int depth = (Integer) input;
        int result = 0;
        for (int i = 0; i < 20; i++) {
            result += chainModifications(depth).run(0).state();
        }
        return result;
    });
